extensions/amp-form/0.1/amp-form.js

```diff
@@ -1223,6 +1229,80 @@ export class AmpForm {
     }
   }
 
+  /**
+   * Initialize form fields from query parameter values if attribute
+   * 'data-initialize-from-url' is present on the form and attribute
+   * 'data-allow-initialization' is present on the field.
+   * @private
+   */
+  maybeInitializeFromUrl_() {
+    if (
+      isProxyOrigin(this.win_.location) ||
+      !this.form_.hasAttribute('data-initialize-from-url')
+    ) {
+      return;
+    }
+
+    const valueTags = ['SELECT', 'TEXTAREA'];
+    const valueInputTypes = [
+      'color',
+      'date',
+      'datetime-local',
+      'email',
+      'hidden',
+      'month',
+      'number',
+      'range',
+      'search',
+      'tel',
+      'text',
+      'time',
+      'url',
+      'week',
+    ];
+    const checkedInputTypes = ['checkbox', 'radio'];
+
+    const queryParams = parseQueryString(this.win_.location.search);
+    Object.keys(queryParams).forEach(key => {
+      const field = this.form_./*OK*/ querySelector(`[name="${key}"]`);
+      if (!field) {
+        return;
+      }
+      // Do not interfere with form fields that utilize variable substitutions.
+      // These fields are populated at time of form submission.
+      if (field.hasAttribute('data-amp-replace')) {
+        return;
+      }
+      // Form fields must be whitelisted
+      if (!field.hasAttribute('data-allow-initialization')) {
+        return;
+      }
+
+      const value = queryParams[key] || '';
+      const type = field.getAttribute('type') || 'text';
+      const tag = field.tagName;
+
```

Comment:
What do you think about using `isFieldDefault` and `isFieldEmpty` methods from the `src/form` to early return if the user has already interacted with the input? That way we also don't have to worry about the field value checks before setting them. This suggestion assumes that the form would not want to change/clear any user progress that has been made.

Reply:
My original intention with the field value checks was to determine whether to trigger a form update event after all fields were updated. Since that feature is removed (we settled on the form being considered clean instead of dirty after populating from URL), I suppose these checks are not strictly necessary. I still feel uneasy about setting `.value` when unnecessary (thinking about additional listeners that may be attached to inputs in the future), so I tend to like the field value checks.

W.r.t `isFieldDefault` and `isFieldEmpty`, do you think it's a significant concern that a user could modify a form before `amp-form.js` loads and the constructor runs?

Reply:
Based on how the generic form dirtiness check is implemented it looks like it is really only a concern at initialization if amp-bind has modified the inputs but not if the user has. In that case I think this is fine as is.

Reply:
I appreciate you taking the time to look through the PR closely. Thanks much for the review.